### PR TITLE
add cache to jwkset (public keys) and jwk uri

### PR DESCRIPTION
**Pull request by bendaamerahm**
- add cache to public keys
- add retry after fail on validation from cache data

### CODE REVIEW

File: src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java [hunk 1]
```diff
@@ -109,29 +125,71 @@ public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
...
-                        exchange.getRequest()
-                                .mutate()
-                                .headers(h -> h.set(HEADER_USER_ID, jwtClaimsSet.getSubject()));
-
-                        return chain.filter(exchange);
-                    } catch (JOSEException | BadJOSEException | MalformedURLException e) {
-                        LOGGER.info("{}: 401 Unauthorized, The token cannot be trusted : {}", exchange.getRequest().getPath(), e.getMessage());
-                        return completeWithCode(exchange, HttpStatus.UNAUTHORIZED);
-                    }
+        return jwkUriCache.get(iss.getValue()) != null ? proceedFilter(new FilterInfos(exchange, chain, jwt, jwtClaimsSet, iss, clientID, jwsAlg, jwkUriCache.get(iss.getValue()))) : gatewayService.getJwksUrl(jwtClaimsSet.getIssuer())
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java:128]: I'd prefere not searching twice, so I'd go with a local variable, even though Jamal may not approve ;-)
And here, as either branch is wordy, I'd go with a 'if' in stead of a ternary.
And may be yet, leave old return as it was, just "intercepting" if the cache contains what we are looking for.

File: src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java [hunk 2]
```diff
@@ -109,29 +125,71 @@ public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
...
+                jwkUriCache.remove(filterInfos.getIss().getValue());
+                LOGGER.info(UNAUTHORIZED_INVALID_PLAIN_JOSE_OBJECT_ENCODING, filterInfos.getExchange().getRequest().getPath());
+                return completeWithCode(filterInfos.getExchange(), HttpStatus.UNAUTHORIZED);
+            }
+        } else {
+            try {
+                validate(filterInfos);
+            } catch (BadJOSEException | JOSEException e) {
+                jwkSetCache.remove(filterInfos.getIss().getValue());
+                this.proceedFilter(new FilterInfos(filterInfos.getExchange(), filterInfos.getChain(), filterInfos.getJwt(), filterInfos.getJwtClaimsSet(), filterInfos.getIss(), filterInfos.getClientID(), filterInfos.getJwsAlg(), filterInfos.getJwkSetUri()));
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java:175]: Why do we create an exact copy ?

File: src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java [hunk 3]
```diff
@@ -109,29 +125,71 @@ public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
...
+            }
+        }
+
+        return filterInfos.getChain().filter(filterInfos.getExchange());
+    }
+
+    @AllArgsConstructor
+    @Getter
+    @Setter
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java:184]: Are the setters needed ?

**Comment by SlimaneAmar** [comment]
Where are all my comments ?

File: src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java (hunk 1 above)
Comment (bendaamerahm) [reply src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java:121]: ill change the code to use local variable unstead

File: src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java (hunk 2 above)
Comment (bendaamerahm) [reply src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java]: my bad , ill fix it

File: src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java (hunk 3 above)
Comment (bendaamerahm) [reply src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java:179]: nice catch , ill fix it

**Review by laurgarn — APPROVED** [review]
Looks good

**Review by jonenst — APPROVED** [review]

**Pull request merged**